### PR TITLE
feat(cb2-5728): auto focus date segments

### PR DESCRIPTION
**Pull request by mcadenas-bjss**
## Date Component Improvement

_Automate the focus event between date segments as the user types numeric values in._
[link to ticket number]()

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/forms/components/date/focus-next.directive.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      }
+    } else if (value.length === 4 && this.includeTime && 'year' === currentSegment) {
+      this.focus(`${name}-hour`);
+      return;
+    }
+  }
+
+  focus(el: string) {
+    document.getElementById(el)?.focus();
+  }
```
Comment (tahayazici) [inline src/app/forms/components/date/focus-next.directive.ts:39]: As requested, here is an alternative in which calculating the suffix is abstracted:
```suggestion
    const segments = id.split('-');

    const suffix = this.getSuffix(value, segments.splice(-1)[0]);

    if (suffix) {
      document.getElementById(segments.join('-') + suffix)?.focus();
    }
  }

  getSuffix(value: any, currentSegment: any): string | undefined {
    if (value.length === 2) {
      if ('day' === currentSegment) {
        return '-month';
      } else if ('month' === currentSegment) {
        return '-year';
      } else if ('hour' === currentSegment) {
        return '-minute';
      }
    } else if (value.length === 4 && this.includeTime && currentSegment === 'year') {
      return '-hour';
    }
  }
```

Comment (tahayazici) [inline src/app/forms/components/date/focus-next.directive.ts:39]: As requested, here is an alternative in which calculating the suffix is abstracted:
```suggestion
  @HostListener('input', ['$event'])
  onInput(e: InputEvent) {
    const segments = this.el.nativeElement.id.split('-');

    const suffix = this.getSuffix(segments.splice(-1)[0]);

    if (suffix) {
      document.getElementById(segments.join('-') + suffix)?.focus();
    }
  }

  getSuffix(currentSegment: any): string | undefined {
    const { nativeElement: value } = this.el;

    if (value.length === 2) {
      if ('day' === currentSegment) {
        return '-month';
      } else if ('month' === currentSegment) {
        return '-year';
      } else if ('hour' === currentSegment) {
        return '-minute';
      }
    } else if (value.length === 4 && this.includeTime && currentSegment === 'year') {
      return '-hour';
    }
  }
```

Comment (mcadenas-bjss) [reply src/app/forms/components/date/focus-next.directive.ts]: Thank, good idea! I've implemented it with some small changes.

File: src/app/forms/components/date/date.component.html [hunk 2]
```diff
@@ -10,9 +10,11 @@
           <input
             #dayEl="ngModel"
             appNumberOnly
+            appFocusNext
+            [includeTime]="includeTime"
             class="govuk-input govuk-date-input__input govuk-input--width-2"
             [ngClass]="{ 'govuk-input--error': elementHasErrors(0) && (dayEl.dirty || dayEl.touched) }"
-            id="{{ name }}-day"
+            attr.id="{{ name }}-day"
```
Comment (gjulien-bjss) [inline src/app/forms/components/date/date.component.html:17]: Why the change? especially since the other components don't have that change?
```suggestion
            id="{{ name }}-day"
```

**Review by gjulien-bjss — APPROVED** [review]

**Review by kpearmanbjss — APPROVED** [review]
